aptos-move/framework/aptos-framework/sources/configs/Stake.move

```diff
@@ -0,0 +1,247 @@
+module AptosFramework::Stake {
+    use Std::Vector;
+    use Std::Signer;
+    use CoreFramework::SystemAddresses;
+    use CoreFramework::Timestamp;
+//    use CoreFramework::Reconfiguration;
+    use AptosFramework::TestCoin::{Self, Coin};
+
+    const MINIMUM_LOCK_PERIOD: u64 = 86400;
+    const MINIMUM_RECONFIG_PERIOD: u64 = 3600;
+
+    /// Basic unit of stake delegation, it's stored in StakeInfo.
+    struct Delegation has store {
+        coin: Coin,
+        from: address,
+        locked_until_secs: u64,
+    }
+
+    /// Aggregation of delegation and represent a validator's voting power, stored in ValidatorInfo.
+    /// Invariants:
+    /// 1. current_stake = sum(active + cooldown)
+    /// 2. user interact with warmup and inactive if it's in the ValidatorSet.
+    /// 3. user interact with active, inactive if it's not in the ValidatorSet.
+    /// 4. warmup and cooldown are empty if it's not in the ValidatorSet.
+    struct StakeInfo has store {
+        // sum of active and cooldown stakes, updated on epoch boundary.
+        current_stake: u64,
+        // active stake
+        active: vector<Delegation>,
+        // inactive stake, can be withdrawn
+        inactive: vector<Delegation>,
+        // pending activation for next epoch
+        warmup: vector<Delegation>,
+        // pending deactivation for next epoch
+        cooldown: vector<Delegation>,
+    }
+
+    /// Consensus information per validator, stored in validator address.
+    struct ValidatorInfo has key {
+        stake_info: StakeInfo,
+        consensus_pubkey: vector<u8>,
+        network_address: vector<u8>,
+    }
+
+    /// Full ValidatorSet, stored in @CoreResource.
+    struct ValidatorSet has key {
+        consensus_scheme: u8,
+        // minimum stakes requires to join validator set
+        minimum_stake: u64,
+        // maximum stakes allowed to join validator set
+        maximum_stake: u64,
+        // should we store the ValidatorInfo in place?
+        validators: vector<address>,
+        last_update_time_secs: u64,
+    }
+
+    /// Any user can delegate a stake.
+    fun delegate_stake(account: &signer, to: address, coin: Coin, locked_until_secs: u64) acquires ValidatorInfo, ValidatorSet {
+        let current_time = Timestamp::now_seconds();
+        assert!(current_time + MINIMUM_LOCK_PERIOD < locked_until_secs, 0);
+        let stake_info = &mut borrow_global_mut<ValidatorInfo>(to).stake_info;
+        let delegation = Delegation {
+            coin,
+            locked_until_secs,
+            from: Signer::address_of(account),
+        };
+        // add to warmup if it's a current validator otherwise add to active directly
+        if (is_validator(to)) {
+            Vector::push_back(&mut stake_info.warmup, delegation);
+        } else {
+            stake_info.current_stake = stake_info.current_stake + TestCoin::value(&delegation.coin);
+            Vector::push_back(&mut stake_info.active, delegation);
+        }
```

Comment:
it seems weird to directly add to a running validator as this will cause changes in vote weight in the middle of an epoch. I suspect there might be some easier computations by keeping stake steady throughout an epoch.

Reply:
no, it doesn't change the weight, the weight is sum(active, cooldown), not warmup